### PR TITLE
Don't automatically resubmit jobs when they fail to provision

### PR DESCRIPTION
**Pull request by plars**

### CODE REVIEW

File: testflinger_agent/agent.py [hunk 1]
```diff
@@ -165,10 +165,6 @@ def process_jobs(self):
                     # In this case, we need to mark the device offline
                     if exitcode == 46:
                         self.mark_device_offline()
-                        self.client.repost_job(job_data)
-                        shutil.rmtree(rundir)
-                        # Return NOW so we don't keep trying to process jobs
-                        return
```
Comment (nadzyah) [inline testflinger_agent/agent.py:171]: Don't we want to exit from the function here (return)?

**Comment by plars** [comment]
tl;dr - I don't think the automatic resubmit is really helpful, and sometimes results in unexpected errors and job states. Let's disable it.

The long version for you masochists who want all the gory details:
This could be a little controversial, but I hope not. We always avoided automatically "retrying" a failed job for *every* provisioning failure, because in a pathological case, you could have a bad image that bricks every single device of that type. However, some device agents are able to detect a few (but not all) cases where the reason provisioning failed is because the device is broken and not something that could be wrong with the image. IFF we determine that it's a broken device (ex. we can't reach the muxpi controller, or maas is unable to release the system, for example), then we will sometimes try to just resubmit the job back to the same queue under the same job id. That way you can still track it and have a "bonus chance" to get it to run without having to resubmit it.

I think this was clever, but not really all that helpful. Because often times, it's the *only* agent of that type. If you aren't paying attention, then the first run marks the agent offline because something is broken, resubmits it, then all you see is that your job is still waiting and you have no idea why.

However, now that we have things like multi-device jobs, a web ui, etc... things get really weird. IIRC, Marian saw this problem with subordinate jobs in a multi-device setup that failed the run because they failed to provision (this is correct) but then resubmitted themselves. So when he tried to rerun the multi-job, he had a bad situation there. Also, saviq hit a weird corner case where a job got resubmitted, hit the 503 error, later retried but then couldn't reach the server again, and went away but now there are two result blobs for it on the job, one of which thinks it's still running.

File: testflinger_agent/agent.py (hunk 1 above)
Comment (plars) [reply testflinger_agent/agent.py:171]: No, I'd actually like it to fall through and finish out submitting the results it has so far, which may be useful for debugging. That way the requester will see the message that it failed to provision and any other info gathered so far.

**Review by nadzyah — APPROVED** [review]

**Comment by jocave** [comment]
I think this would be a good time to capture what the steps that will occur during a job are in documentation - this is notable change in behaviour that I think users should be able to understand without following code changes or being notified directly by us.

I think maybe an explanation doc would be appropriate? Given the separation in repos that would be a commit to the testflinger project.

**Comment by plars** [comment]
> I think this would be a good time to capture what the steps that will occur during a job are in documentation - this is notable change in behaviour that I think users should be able to understand without following code changes or being notified directly by us.
> 
> I think maybe an explanation doc would be appropriate? Given the separation in repos that would be a commit to the testflinger project.

Is https://github.com/canonical/testflinger-agent/blob/main/README.rst#test-phases what you're looking for or something else?

**Comment by hum4n0id** [comment]
Latest changes look good from my end.

**Review by hum4n0id — APPROVED** [review]
